Added qid version of Named Qubit

oliverob opened the pull request:
closes #3069

I have added a qid version of Named Qubit, and added the relevant tests

oliverob commented:
@googlebot I signed it!

cduck commented:
Thanks for implementing this.  Can you remove all the formatting changes to code unrelated to implementing NamedQid?  This adds a lot of noise making it hard to review.  If the auto-format did this to the entire file, then it's a bug; auto-format should only reformat changed lines.  Also, please use single quotes `'...'` for all strings (except doc strings or strings containing a literal `'`) to stay consistent with the style of the rest of Cirq.

It looks like you have implemented everything needed for NamedQid.  I can give feedback on any smaller issues once you fix the above.

oliverob commented:
I have removed the formatting changes. They were added by my autoformatter. I have also changed the double quotes to single quotes. Hope that is all good now.

balopat commented on cirq/ops/named_qubit_test.py:
```diff
@@ -1,13 +1,13 @@
 # Copyright 2018 The Cirq Developers
 #
-# Licensed under the Apache License, Version 2.0 (the "License");
+# Licensed under the Apache License, Version 2.0 (the 'License');
```
keep double quote

balopat commented on cirq/ops/named_qubit_test.py:
```diff
@@ -1,13 +1,13 @@
 #     https://www.apache.org/licenses/LICENSE-2.0
 #
 # Unless required by applicable law or agreed to in writing, software
-# distributed under the License is distributed on an "AS IS" BASIS,
```
keep double quotes

balopat commented on cirq/ops/named_qubit.py:
```diff
@@ -33,29 +37,112 @@ def __init__(self, name: str) -> None:
 
-        NamedQubit.range(3, prefix="a") -> ["a1", "a2", "a3]
-        NamedQubit.range(2, 4, prefix="a") -> ["a2", "a3]
+        NamedQubit.range(3, prefix='a') -> ['a1', 'a2', 'a3']
```
keep double quotes

balopat replied:
Good question. I might have jumped the gun here without a solid foundation, as I thought it was an unnecessary find/replace from double to single. Let me find guidance on this.

balopat replied:
The only guidance in https://google.github.io/styleguide/pyguide.html is to stay consistent within a file: 

> Be consistent with your choice of string quote character within a file. Pick ' or " and stick with it. It is okay to use the other quote character on a string to avoid the need to \\ escape within the string.

cduck commented on cirq/protocols/json_serialization.py:
```diff
@@ -157,7 +158,7 @@ def two_qubit_matrix_gate(matrix):
                 'VirtualTag': cirq.VirtualTag,
                 'WaitGate': cirq.WaitGate,
-                '_QubitAsQid': raw_types._QubitAsQid,
+                '_QubitAsQid': cirq.NamedQid,
```
This is incorrect.  Instead of changing `_QubitAsQid` to `NamedQid`, change the test to another type of qubit that doesn't have a qudit variant.

cduck commented on cirq/protocols/json_test_data/NamedQid.repr:
```diff
@@ -0,0 +1 @@
+cirq.NamedQid("a",dimension=3)
```
Nit:
```suggestion
cirq.NamedQid("a", dimension=3)
```

cduck commented on cirq/protocols/json_test_data/_QubitAsQid.json:
```diff
@@ -1,8 +1,5 @@
-    "cirq_type": "NamedQubit",
-    "name": "a"
-  },
+  "cirq_type": "NamedQid",
+  "name": "a",
   "dimension": 5
-}
+}
```
Instead of changing _QubitAsQid to NamedQid, change the test to another type of qubit that doesn't have a qudit variant.  Use a custom qubit type defined just for the test if needed.

oliverob commented:
I have made those edits. I wasn't quite sure how to make a custom qubit type to test the _QubitAsQid. My attempt was to create a new class called NoIdentifierQubit.

cduck commented:
Sorry for not being clear.  For NoIdentifierQubit, it should not be added to the Cirq API.  When a temporary class is needed for a test, just add it to the same *_test.py file and only implement the methods necessary for your test to work.  If it is used by several different tests, then you would consider adding the complete class (with its own tests) under the cirq.testing module.

In this case, to test the json serialization, it may be easier to just test with the repr like `_QubitAsQid(NamedQubit('a'), 3)`.

Another option, since you already implemented NoIdentifierQubit, is just move it into cirq/testing/.

oliverob commented:
Okay I have moved NoIdentifierQubit to cirq/testing/ and I think it is all fine now except for the Doc Test which is failing for reasons I don't understand

cduck commented on cirq/ops/named_qubit.py:
```diff
@@ -33,29 +35,111 @@ def __init__(self, name: str) -> None:
+
+            >>> cirq.NamedQid.range(3, prefix='a', dimension=3)
+            [cirq.NamedQid('a0', dimension=3), cirq.NamedQid('a1', dimension=3)\
+, cirq.NamedQid('a2', dimension=3)]
```
Does it work if you only have 8 spaces here?  I'm worried this will affect the indentation of the rest of the doc string.

cduck commented on cirq/ops/named_qubit_test.py:
```diff
@@ -1,4 +1,4 @@
-# Copyright 2018 The Cirq Developers
+# Copyright 2020 The Cirq Developers
```
Don't change for existing files.

oliverob commented:
Done. I think it should be good to merge now. Thanks so much for all your help!

cduck commented:
Thanks for contributing and your fast responses.

cduck commented:
@balopat The docs build seems to be failing in a way unrelated to the code.  I'm not sure what's going on.

cduck commented:
@oliverob I gave you the wrong advice for documentation code literals.  For reStructuredText, code blocks are surrounded by double backticks (\`\`code\`\`) and text adjacent to the backticks has special meaning so \`\`NamedQid\`\`s should be escaped as \`\`NamedQid\`\`\\s (https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#character-level-inline-markup).

Please fix \`\`NamedQid\`\`\\s or \`NamedQid\`\\s so it passes the build-docs test.  I don't think we are consistent about which we use currently but the double backticks are more correct.

balopat commented:
> @balopat The docs build seems to be failing in a way unrelated to the code. I'm not sure what's going on.

I don't think it is unrelated - 
```
/home/runner/work/Cirq/Cirq/cirq/ops/named_qubit.py:docstring of cirq.NamedQid.range:20: WARNING: Inline interpreted text or phrase reference start-string without end-string.
```

But I see you responded now. 

Thank you both for the great work here!

cduck commented:
Ping @oliverob.  I'd like to get this merged.

It looks like the mypy update caused a new failure.  Making the same change that @Strilanc did a few days ago should be fine (https://github.com/quantumlib/Cirq/commit/aff27b8bd78ac9d2c09f1902372e12f010901eb3#diff-fc9667edd0c0399b27f542ff2b417a2bR26).

Fix the lint error by escaping the new backslashes with another backslash (`\\`).

oliverob commented:
@cduck Sorry it took me so long to get round to it but I have finally made those changes. It won't run the CI test now which is quite odd. There also appears to be a conflicting file which wasn't there before. Not sure if I should try and resolve the conflict as I wouldn't want to do it wrong.

cduck commented:
Someone else probably updated api.rst.  It should be simple to merge it with git.  Run the commands you normally would to update your local branch.  When git says there is a merge conflict, open the file, look for `<<<` and `>>>` which indicate the conflicting versions.  Edit it to have both the correct changes (removing `<<,==,>>`) then rerun the last command with `--continue`.

If the CI doesn't run you can try pushing an empty commit to retry (`git commit --allow-empty`).

balopat commented:
Yeah, the api.rst got moved under `rtd_docs` - it is a bit of an advanced merging scenario, let us know if you get stuck with it. Alternatively you can 
1. revert your api.rst changes
2. merge master 
3. fix the api.rst changes in under rtd_docs now

balopat commented:
@oliverob actually I tried locally and `git merge master` should just work fine. If you [allow us](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to add changes to your branch, we can do this ourselves and merge the PR.

oliverob commented:
Brilliant thanks, can the pull request be merged now? I can't do it as I don't have write accesss. Or has it already been done?
